compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/PropertiesToFieldsLowering.kt

```diff
@@ -72,8 +76,29 @@ class PropertiesToFieldsLowering(val context: CommonBackendContext) : IrElementT
         return accessor.origin == IrDeclarationOrigin.DEFAULT_PROPERTY_ACCESSOR && Visibilities.isPrivate(accessor.visibility)
     }
 
+    private val remappedFieldCache = mutableMapOf<IrField, IrField>()
+
+    private fun IrField.getRealBackingField(): IrField {
+        val irClass = resolveFakeOverride()!!.parent as? IrClass ?: return this
+        if ((irClass.isCompanion || (irClass.isObject && irClass.visibility != Visibilities.LOCAL)) && !isStatic) {
+            // This specific combination of circumstances means it's an imported @JvmField-annotated field.
```

Comment:
I wouldn't really call this a fix, since the hack effectively creates a second orphaned `IrField` for `result`, so the one you arrive at by going through Z's `declarations` member is not the one that is the `symbol.owner` of the `IrGetField`. Removing `fieldReplacementMap` thus would probably be OK, but would increase the number of cases where this anomaly is visible.

Reply:
Right, I've tried to do this, but hadn't finished yet (some tests are still failing). Overall I agree that just removing `fieldReplacementMap` is undesirable at this point. Let me look into the `JvmDeclarationFactory` approach once again after merging your changes. Thanks!

Reply:
If you're interested, I've made my own version of the refactoring: https://github.com/pyos/kotlin/commit/db298f163f6141d5a577e9273999b8d1c76c7f46
The only failing tests are 3 FIR ones because FIR does not set correspondingPropertySymbol for some reason.

Reply:
Yeah, your version looks very good and it's exactly what I had in mind. I suggest you to attach it to this PR then, since it looks very promising (and I'll have some minor comments).

Reply:
Thank you, added the commit here.

Reply:
The FIR problem is fixed now.